src/lib/pubkey/hss_lms/hss_lms_utils.h

```diff
@@ -0,0 +1,79 @@
+/**
+ * Utils for HSS/LMS
+ * (C) 2023 Jack Lloyd
+ *     2023 Fabian Albert, Philippe Lieser - Rohde & Schwarz Cybersecurity GmbH
+ *
+ * Botan is released under the Simplified BSD License (see license.txt)
+ */
+
+#ifndef BOTAN_HSS_LMS_UTILS_H_
+#define BOTAN_HSS_LMS_UTILS_H_
+
+#include <botan/hash.h>
+#include <botan/internal/loadstor.h>
+
+namespace Botan {
+
+/**
+ * @brief Helper class used to derive secret values based in the pseudorandom key generation
+ * described in RFC 8554 Appendix A.
+ *
+ * This generation computes the following:
+ *
+ * Result = Hash( identifier || u32str(q) || u16str(i) || u8str(j) || SEED )
+ *
+ * This Key Generation procedure is also used for the seed derivation function of
+ * SECRET_METHOD 2 defined in https://github.com/cisco/hash-sigs,
+ */
+class PseudorandomKeyGeneration {
+   public:
+      /**
+       * @brief Create a PseudorandomKeyGeneration instance for a fixed @p identifier
+       */
+      PseudorandomKeyGeneration(std::span<const uint8_t> identifier);
+
+      /**
+       * @brief Specify the value for the u32str(q) hash input field
+       */
+      void set_q(uint32_t q) { store_be(q, m_q.data()); }
+
+      /**
+       * @brief Specify the value for the u16str(i) hash input field
+       */
+      void set_i(uint16_t i) { store_be(i, m_i.data()); }
+
+      /**
+       * @brief Specify the value for the u8str(j) hash input field
+       */
+      void set_j(uint8_t j) { *m_j = j; }
```

Comment:
Maybe its worth to hold the locations of `q`, `j` and `i`. But perhaps we should calculate the spans inline? Does this have an impact on the performance?

```suggestion
      void set_q(uint32_t q) { store_be(q, std::span(m_input_buffer).last<7>().first<4>().data()); }

      /**
       * @brief Specify the value for the u16str(i) hash input field
       */
      void set_i(uint16_t i) { store_be(i, std::span(m_input_buffer).last<3>().first<2>().data()); }

      /**
       * @brief Specify the value for the u8str(j) hash input field
       */
      void set_j(uint8_t j) { m_input_buffer.back() = j; }
```

Reply:
At least the `m_j` raw pointer can be avoided like that without a performance impact, for sure.

Reply:
I could not measure a significant performance boost. However, I like getting rid of the member spans.